# evolution-of-syntrophy
Code and data related to analysis of genomics data for syntrophic community.

Sulfate reducing bacteria conditionally engage with methanogenic archaea in interspecies syntrophic interactions that contribute to anaerobic geochemical cycling of over 1 gigatons of C every year.  Prior work has demonstrated that while frequent fluctuations in sulfate availability render a syntrophic community unstable, prolonged laboratory evolution in obligate syntrophy conditions can improve productivity but at the expense of erosion of sulfate respiration (SR) and accumulation of mutations in genes of diverse functions. Hence, fundamental questions remain regarding how a microbial community manages evolutionary tradeoffs between conditional phenotypic switching to syntrophy and increasing the productivity of syntrophic interactions without loss of SR. Here, we have analyzed temporal and combinatorial accumulation of mutations across independent evolution lines of a syntrophic community of Desulfovibrio vulgaris and Methanococcus maripaludis, and dissected at bulk and single cell resolution the functional consequences of interactions among lineages within SR- and SR+ minimal community assemblages from the same parental line. This multiscale analysis has discovered that selection of interactions among specific evolved lineages generates synergistic epistasis that enhances cooperativity within SR- and SR+ assemblages, allowing their co-existence as r- and K-strategists, respectively.

## data:
* SRA Sequence submission files
* Annotated mutations
* Genome fasta files
* single cell annotated mutations

## figures
Manuscript figure illustrator and png files as well as scripts to produce figures when appropriate

## reference
Reference genome files and associated indexes for read alignment to genome

## scripts
* Python scripts for sequencing data analysis and variant calling pipelines
* R-scripts for data processing

## supplementary files
* Supplementary excell files
* Supplementary figure

